ARCHITECTURE.md

```diff
@@ -0,0 +1,196 @@
+# Architecture
+
+This document describes the high-level architecture of the Google Cloud C++
+Client libraries. Its main audience are developers and contributors making
+changes and additions to these libraries, but users of the libraries may also
+find this useful. If you want to familiarize yourself with the code in the
+`google-cloud-cpp` project, you are at the right place.
+
+## What these libraries do
+
+The goal of the libraries is to provide idiomatic C++ libraries to access
+services in [Google Cloud Platform](https://cloud.google.com). All services are
+in scope, though only a handful of services have been implemented to this point.
+
+What do we mean by idiomatic? We mean that C++ developers will find the
+APIs familiar, or "natural", that these APIs will fit well with the rest of the
+C++ ecosystem, that very few new "concepts" are needed to understand how to use
+these libraries.
+
+More specifically, some of the functionality offered by these libraries include:
+
+- The libraries automatically *retry* RPCs that fail with a retriable error,
+  such as network failures.
+- The libraries only retry operations when it is safe to do so, "delete the
+  first row" is not a safe operation to retry when the previous result is
+  unknown. In the documentation we refer to operations that are safe to retry
+  as [idempotent](https://en.wikipedia.org/wiki/Idempotence).
+- The libraries also *resume* large downloads or database reads from the last
+  fully received element, minimizing the amount of duplicate data received by
+  your application.
+- The application can always change the retry and resume policies, the libraries
+  have reasonable defaults, and implement best practices for backing off before
```

Comment:
Perhaps "The libraries have reasonable defaults for retry and resume policies, and implement ...".

Reply:
Took a stab at it.